### ISSUE TITLE
`cargo test` doesn't mention which file(s) in examples/ is affected when the `main` function is missing

### ISSUE THREAD

**Issue by Cobrand** [rust-lang/cargo#4948]
Basically, if you have a lib crate with multiple files in "examples", but *one* of them doesn't have a main, you get this error when running `cargo test`:

```
error[E0601]: main function not found
```

But you don't know which file is affected by this error.

It would be better to have an error message that says "main function not found in examples/something.rs".

Note that it may be related to rustc itself, but just in case I opened an issue here as it seemed more fit.

**Comment by alexcrichton** [comment]
Thanks for the report! Can you gist the full error message as well as perhaps an example project to reproduce?

**Comment by Cobrand** [comment]
Extremely simple example: https://github.com/Cobrand/cargo_error_example

Just `cargo new --lib` and create an empty file in `examples/`.

Full error message:
```
   Compiling cargo_error_example v0.1.0 (file:///home/cobrand/Code/cargo_error_example)
error[E0601]: main function not found

error: aborting due to previous error

error: Could not compile `cargo_error_example`.

To learn more, run the command again with --verbose.
```

**Comment by alexcrichton** [comment]
Hm ok, isn't this covered by 

```
error: Could not compile `cargo_error_example`.
```

?

**Comment by lukaslueg** [comment]
The error refers to the name of the package, not the name of the example. Put a file `examples/foo.rs` and the error still says `cargo_error_example`.

There are three things going on here:

* We could place an additional (?) warning [here](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_rustc/job_queue.rs#L232) and refer to `key.target.src_path().to_str_lossy()` or something.
* The errors in part come directly from `rustc` and `cargo` itself and the output is already quite clobbered. Notice that there are four `error:`-lines and one `warning:`-line referring to the same error.
* The overall output is actually different depending on wether or not the queue is empty. Running `cargo test` the first time (or after `cargo clean`) gives

```
Compiling issue4948 v0.1.0 (file:///home/lukas/issue4948)
error[E0601]: main function not found

error: aborting due to previous error

error: Could not compile `issue4948`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

Running a second time gives
```
Compiling issue4948 v0.1.0 (file:///home/lukas/issue4948)
error[E0601]: main function not found

error: aborting due to previous error

error: Could not compile `issue4948`.

To learn more, run the command again with --verbose.
```

Notice that there is no warning, no `error: build failed`, yet a `info` referring to `--verbose` the second time around. Maybe this needs to get cleaned up?

**Comment by ehuss** [comment]
One option is to just change the error message ([here](https://github.com/rust-lang/rust/blob/da569fa9ddf8369a9809184d43c600dc06bd4b4d/src/librustc/middle/entry.rs#L165)) to include the crate name.  Messages without spans (thus filenames) are very rare, and this is the only one that I think might be confusing.  Another example is the lint warning that a crate name should have snake case, but it clearly tells you the name of the crate.  Other messages without spans include linker errors and incorrect command-line arguments (presumably the user can figure these out?).

**Comment by ehuss** [comment]
I think this can be closed now that #48706 has been merged.

**Comment by alexcrichton** [comment]
Great!
